aead/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+//! A set of traits designed to support authenticated encryption.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(all(feature = "alloc", not(has_extern_crate_alloc)), feature(alloc))]
+
+#![cfg(feature = "std")]
+extern crate core;
+extern crate generic_array;
+
+use generic_array::typenum::Unsigned;
+use generic_array::{GenericArray, ArrayLength};
+use core::marker::PhantomData;
+
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct Error;
+
+/// A trait which can support a stateful, RFC5116 authenticated encryption
+/// scheme with a fixed-size nonce.
+pub trait Aead {
+    /// The key size in a new method.
+    type KeySize: ArrayLength<u8> + Unsigned;
+    /// The length of a nonce.
+    type NonceSize: ArrayLength<u8> + Unsigned;
+    /// The maximum length of the nonce.
+    type TagSize: ArrayLength<u8> + Unsigned;
+    /// The amount of suffix padding, in bytes, which needs to be appended to
+    /// a plaintext to accommodate this cipher's output.
+    type CiphertextOverhead: ArrayLength<u8> + Unsigned;
+
+    /// Construct a new stateful instance for the given key.
+    fn for_key(key: GenericArray<u8, Self::KeySize>) -> Self;
+
+    /// Perform an in-place encryption of the given plaintext, which is built
+    /// from the first plaintext_used bytes of the pre-populated plaintext
+    /// buffer.
+    ///
+    /// Implementers are responsible for shifting any existing contents of the
+    /// plaintext, if necessary, and returning a slice trimmed to the
+    /// algorithm-specific ciphertext.
```

Comment:
The implementation of the trait for this will need to do the `rotate_right()` for itself.

Reply:
So with such scheme first N bits will get silently overwritten? I think it's a very error-prone design, especially in generic contexts.

Reply:
No, you've got a plaintext like:

`plaintext = "This is my plaintext.\0\0\0\0\0\0\0\0\0", used = 21`

And the algorithm needs the first 8 bytes for itself, it does `plaintext.rotate_right(8)`, the resulting buffer is now:

`plaintext = "\0\0\0\0\0\0\0\0This is my plaintext.\0"`

Algorithms will need to check that `plaintext.len() - used > CiphertextOverhead`.

Reply:
Ah, I understand. As I've said earlier, personally I don't quite like performance implications of such approach.

@tarcieri 
What do you think? Does it worth to bound performance in this way for API simplification?

Reply:
My preferred solution here, at least for now, is to spec a detached API and then circle back on how to do a non-allocating attached API in a separate PR.

I have some ideas about how to do it, but I'd rather discuss them in isolation, and done correctly the allocating, non-allocating attached, and non-allocating detached APIs can each build on the next.

Reply:
I think the detached API you suggest would need to take header buffers, trailer buffers, and return slices from both, otherwise there is an incompatibility risk with standards that use variable-length headers/trailers.

Reply:
Well, really `draft-mcgrew-aead-aes-cbc-hmac-sha2` is the only thing that doesn't work well with a detached API since it's using a block cipher instead of a stream cipher and therefore needs the extra overhead for the block padding. A detached API punts on headers/trailers and will work just fine with AES-SIV's prefix tags if we defer assembling the message to the allocating API for now.

I have a more concrete suggestion which I'll make not as a line note.